File: V&F Basket Backend/src/main/java/com/V/FBasket/VnFBasket/service/ProductService.java
package com.V.FBasket.VnFBasket.service;

import java.io.IOException;
import java.util.List;

import com.V.FBasket.VnFBasket.dto.ProductRequest;
import com.V.FBasket.VnFBasket.dto.ProductResponseDTO;
import com.V.FBasket.VnFBasket.model.Products;
import org.springframework.web.multipart.MultipartFile;

public interface ProductService {
    Products addProducts(ProductRequest productRequest, MultipartFile file) throws IOException;
    List<ProductResponseDTO> getAllProducts();
    ProductResponseDTO getProductByProductName(String productName);
    List<Products> getProductsByCategoryId(Long categoryId);
    ProductResponseDTO getProductById(Long productId);
    Products updateProduct(ProductRequest product) throws  IOException;
    boolean deleteProduct(Long productId);
    List<ProductResponseDTO> getProductsByCategoryName(String categoryName);

    
} 

File: V&F Basket Frontend/VnFBasketUI/src/app/services/product-service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable, signal } from '@angular/core';
import { Product } from '../models/product.model';
import { BehaviorSubject, Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private apiUrl: string = 'http://localhost:8080/vnfbasket';
  private http = inject(HttpClient);
   private searchText = new BehaviorSubject<string>('');
  searchText$ = this.searchText.asObservable();

  setSearchText(text: string) {
    this.searchText.next(text);
  }

  addProducts(productformData: FormData): Observable<Product>{
    return this.http.post<Product>(this.apiUrl + '/addProduct', productformData);
  }

  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(this.apiUrl + '/getAllProducts');
  }

  getProductsByCategory(category: string): Observable<Product[]> {
    return this.http.get<Product[]>(this.apiUrl + '/getProductsByCategoryName?categoryName=' + category);
  }
  
  getProductByName(productName: string): Observable<Product> {
    return this.http.get<Product>(this.apiUrl + '/getProductByProductName/' + productName);
  }

  getProductById(productId: number): Observable<Product> {
    return this.http.get<Product>(this.apiUrl + '/getProductsById/' + productId);
  }

  deleteProductById(productId: number): Observable<void> {
    return this.http.delete<void>(this.apiUrl+'/deleteProduct/'+productId);
  }

  updateProduct(formData: FormData) {
  return this.http.put(this.apiUrl+'/updateProduct', formData);
}
}
